### PR TITLE
fixes #26887 - graphql login: count bruteforce attempts

### PR DESCRIPTION
**Pull request by timogoebel**

### CODE REVIEW

**Comment by timogoebel** [comment]
Test failures related. :-/

**Comment by timogoebel** [comment]
Test failures unrelated.

@tbrisker: We should have this in 1.23 GA as the current code allows a user to bypass the bruteforce protection.

**Comment by ares** [comment]
Is this a good candidate for 1.22.1 as well?

**Comment by timogoebel** [comment]
@ares: This is a follow up for #6627, which is only in 1.23. So we should just make sure this is in 1.23 as well.

File: app/controllers/concerns/foreman/controller/bruteforce_protection.rb [hunk 1]
```diff
@@ -1,20 +1,19 @@
...
   extend ActiveSupport::Concern
 
-  def count_login_failure
-    Rails.cache.write("failed_login_#{request.ip}", get_login_failures + 1, :expires_in => 5.minutes)
-  end
-
-  def get_login_failures
-    Rails.cache.fetch("failed_login_#{request.ip}") {0} if request.ip.present?
+  included do
+    delegate :count_login_failure, :get_login_failures, :bruteforce_attempt?, :log_bruteforce, to: :bruteforce_protection
```
Comment (timogoebel) [inline app/controllers/concerns/foreman/controller/bruteforce_protection.rb]: Note to self: this has bruteforce_attempt? which is overridden below and should be removed.

**Comment by ares** [comment]
Ah got it, thanks!

**Comment by timogoebel** [comment]
@ares, @tbrisker, @xprazak2: We need this in 1.23 for security reasons. Mind taking a look?

File: app/graphql/mutations/sign_in_user.rb [hunk 2]
```diff
@@ -13,8 +13,22 @@ class SignInUser < BaseMutation
     def resolve(username:, password:)
       return unless username && password
 
+      bruteforce_protection = Foreman::BruteforceProtection.new(
+        request_ip: context[:request_ip]
+      )
+
+      if bruteforce_protection.bruteforce_attempt?
+        bruteforce_protection.log_bruteforce
+        raise GraphQL::ExecutionError, _('Too many tries, please try again in a few minutes.')
```
Comment (ares) [inline app/graphql/mutations/sign_in_user.rb:22]: should this be translated or just extracted? If it's an exception user will be facing, maybe the exception class should inherit from Foreman::Exception (I didn't check if it isn't the case already)

**Review by ares — COMMENTED** [review]
Sorry, I don't have time for full testing, I'm off next week, but looks good code wise, just one question re exception

File: app/graphql/mutations/sign_in_user.rb (hunk 2 above)
Comment (timogoebel) [reply app/graphql/mutations/sign_in_user.rb:22]: I think we're good here.

> If a field's resolve function returns a ExecutionError, the error will be inserted into the response's `“errors”` key and the field will resolve to `nil`.(https://www.rubydoc.info/github/rmosolgo/graphql-ruby/GraphQL/ExecutionError)

**Review by tbrisker — APPROVED** [review]
Thanks @timogoebel !

**Pull request merged**